diff --git a/Application/mt/mt_sys.c b/Application/mt/mt_sys.c
index 279a97f..d58ca3a 100644
--- a/Application/mt/mt_sys.c
+++ b/Application/mt/mt_sys.c
@@ -514,9 +514,10 @@ static void MT_SysVersion(void)
   uint8 *pBuf = &verStr[sizeof(MTVersionString)];
 #if (defined MAKE_CRC_SHDW) || (defined FAKE_CRC_SHDW)  //built for bootloader
   uint32 sblSig;
-  uint32 sblRev;
 #endif
 
+  uint32 sblRev;
+
   MAP_osal_memcpy(verStr, (uint8 *)MTVersionString, sizeof(MTVersionString));
 
 #if (defined MAKE_CRC_SHDW) || (defined FAKE_CRC_SHDW)  //built for bootloader
diff --git a/Application/mt/mt_version.c b/Application/mt/mt_version.c
index 305dfc7..51d8890 100644
--- a/Application/mt/mt_version.c
+++ b/Application/mt/mt_version.c
@@ -69,7 +69,7 @@
  *****************************************************************************/
 const uint8 MTVersionString[] = {
                                    2,  /* Transport protocol revision */
-                                   0,  /* Product ID */
+                                   1,  /* Product ID */
                                    2,  /* Software major release number */
                                    7,  /* Software minor release number */
                                    1,  /* Software maintenance release number */
diff --git a/Application/mt/revision_info.h b/Application/mt/revision_info.h
new file mode 100644
index 0000000..0eb55ad
--- /dev/null
+++ b/Application/mt/revision_info.h
@@ -0,0 +1,4 @@
+#ifndef APPLICATION_MT_REVISION_INFO_H_
+#define APPLICATION_MT_REVISION_INFO_H_
+#define CODE_REVISION_NUMBER 20190515
+#endif
\ No newline at end of file
diff --git a/Stack/Config/preinclude.h b/Stack/Config/preinclude.h
new file mode 100644
index 0000000..e6436c2
--- /dev/null
+++ b/Stack/Config/preinclude.h
@@ -0,0 +1,3 @@
+#define INCLUDE_REVISION_INFORMATION
+#define MT_SYS_KEY_MANAGEMENT 1
+#define MULTICAST_ENABLED FALSE
diff --git a/Stack/Config/znp_cnf.opts b/Stack/Config/znp_cnf.opts
index 1be73a8..81f5f4e 100644
--- a/Stack/Config/znp_cnf.opts
+++ b/Stack/Config/znp_cnf.opts
@@ -93,3 +93,5 @@
 
 /* BDB interface */
 -DMT_APP_CNF_FUNC
+
+--preinclude=preinclude.h
